File: C++/938.cpp
__________________________________________________________________________________________________
sample 96 ms submission
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int sum=0;
    void inorder(TreeNode* r, int L, int R){
        if(r == NULL)
            return;
        if(L <= r->val)
            inorder(r->left, L, R);
        if((L <= r->val)&&(R >= r->val))
            sum += r->val;
        if(R >= r->val)
            inorder(r->right, L, R);
            
    }
    int rangeSumBST(TreeNode* root, int L, int R) {
        ios_base::sync_with_stdio(false);
        cin.tie(NULL);
        inorder(root, L, R);
        return sum;
    }
};
__________________________________________________________________________________________________
sample 41008 kb submission
static const auto fastIO = []() {
    std::ios_base::sync_with_stdio(false);
    std::cin.tie(0);
    std::cout.tie(0);
    return 0;
} ();

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int rangeSumBST(TreeNode* root, int L, int R) {
        if (!root) return 0;
        
        int left = rangeSumBST(root->left, L, R);
        int right = rangeSumBST(root->right, L, R);
        return left + right + ((root->val >= L && root->val <= R) ? root->val : 0);
    }
};
__________________________________________________________________________________________________

File: Java/938.java
__________________________________________________________________________________________________
sample 0 ms submission
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public int rangeSumBST(TreeNode root, int L, int R) {
        int sum=0;
        
        if(root==null){
            return 0;
        }
        
        else{
           // int sum=0;
            if(L<=root.val && R>=root.val){
                sum+=root.val;
            }
            if(L<root.val){
                sum+=rangeSumBST(root.left,L,R);
            }
            if(R>root.val){
                sum+=rangeSumBST(root.right,L,R);
            }
        }
        return sum;
    }
}
__________________________________________________________________________________________________
sample 43388 kb submission
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    int sum;
    public int rangeSumBST(TreeNode root, int L, int R) {
        sum = 0;
        // rangeSumHelper(root, L, R);
        Queue<TreeNode> q = new LinkedList();
        q.add(root);
        while(!q.isEmpty()){
            TreeNode n = q.poll();
            if(n != null){
                if(n.val >= L && n.val <= R){
                    sum += n.val;
                }

                if(n.val > L){
                    q.add(n.left);
                }
                if(n.val < R){
                    q.add(n.right);
                }
            }
        }
        return sum;
    }
}
__________________________________________________________________________________________________
